tests/php/Core/ClassInfoTest.php

```diff
@@ -189,4 +199,67 @@ public function testDataClassesFor()
         ClassInfo::reset_db_cache();
         $this->assertEquals($expect, ClassInfo::dataClassesFor(strtolower($classes[2])));
     }
+
+    /**
+     * @covers \SilverStripe\Core\ClassInfo::classesWithExtension()
+     */
+    public function testClassesWithExtension()
+    {
+        // test existing config based extensions
+
+        $expected = [
+            'silverstripe\\core\\tests\\classinfotest\\extendtest' => ExtendTest::class,
+            'silverstripe\\core\\tests\\classinfotest\\extendtest2' => ExtendTest2::class,
+            'silverstripe\\core\\tests\\classinfotest\\extendtest3' => ExtendTest3::class,
+        ];
+        $this->assertEquals(
+            $expected,
+            ClassInfo::classesWithExtension(ExtensionTest1::class, BaseObject::class),
+            'ClassInfo::testClassesWithExtension() returns class with extensions applied via class config'
+        );
+
+        $expected = [
+            'silverstripe\\core\\tests\\classinfotest\\extendtest' => ExtendTest::class,
+            'silverstripe\\core\\tests\\classinfotest\\extendtest2' => ExtendTest2::class,
+            'silverstripe\\core\\tests\\classinfotest\\extendtest3' => ExtendTest3::class,
+        ];
+        $this->assertEquals(
+            $expected,
+            ClassInfo::classesWithExtension(ExtensionTest1::class, ExtendTest::class, true),
+            'ClassInfo::testClassesWithExtension() returns class with extensions applied via class config, including the base class'
+        );
+
+        // dynamically add extension
+
+        $expected = [];
+        $this->assertEquals(
+            $expected,
+            ClassInfo::classesWithExtension(ExtensionTest2::class, BaseObject::class),
+            'ClassInfo::testClassesWithExtension() returns no classes for extension that hasn\'t been applied yet.'
+        );
+
+        ExtendTest::add_extension(ExtensionTest2::class);
+
+        $expected = [
+            'silverstripe\\core\\tests\\classinfotest\\extendtest2' => ExtendTest2::class,
+            'silverstripe\\core\\tests\\classinfotest\\extendtest3' => ExtendTest3::class,
+        ];
+        $this->assertEquals(
+            $expected,
+            ClassInfo::classesWithExtension(ExtensionTest2::class, ExtendTest::class),
+            'ClassInfo::testClassesWithExtension() returns class with extra extension dynamically added'
+        );
+
+
+        // dynamically remove extension
+
+        ExtendTest::remove_extension(ExtensionTest1::class);
+
+        $expected = [];
+        $this->assertEquals(
+            $expected,
+            ClassInfo::classesWithExtension(ExtensionTest1::class, BaseObject::class),
+            'ClassInfo::testClassesWithExtension() returns no classes after an extension being removed'
+        );
```

Comment:
Would you mind splitting this test up into separate methods? I.e. `testClassesWithExtensionUsingConfiguredExtensions`, `testClassesWithExtensionUsingDynamicallyAddedExtensions` and `testClassesWithExtensionWithDynamicallyRemovedExtensions`

Reply:
Done